File: 2025/23. Merge k Sorted Lists/Solution.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode mergeKLists(ListNode[] lists) {
        Queue<ListNode> heap = new PriorityQueue<>((x, y) -> x.val - y.val);
        ListNode head = new ListNode(), dummy = head;
        for (int i = 0; i < lists.length; i++) {
            if (lists[i] != null)
                heap.add(lists[i]);
        }
        while (!heap.isEmpty()) {
            ListNode t = heap.remove();
            if (t.next != null) {
                heap.add(t.next);
            }
            head.next = t;
            head = head.next;
        }
        return dummy.next;
    }
}

File: 2025/23. Merge k Sorted Lists/solution.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class HeapNode:
    def __init__(self, node):
        self.node = node

    def __lt__(self, other):
        return self.node.val < other.node.val

class Solution:
    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
        heap = [HeapNode(node) for node in lists if node is not None]
        heapq.heapify(heap)
        dummy = head = ListNode()
        while heap:
            node = heapq.heappop(heap).node
            if node.next:
                heapq.heappush(heap, HeapNode(node.next))
                node.next = None
            head.next = node
            head = head.next
        return dummy.next